File: src/main/java/me/alegian/thavma/impl/client/screen/WorkbenchScreen.java
package me.alegian.thavma.impl.client.screen;

import me.alegian.thavma.impl.client.T7GuiGraphics;
import me.alegian.thavma.impl.client.renderer.AspectRenderer;
import me.alegian.thavma.impl.client.texture.Texture;
import me.alegian.thavma.impl.common.aspect.AspectStack;
import me.alegian.thavma.impl.common.menu.WorkbenchMenu;
import me.alegian.thavma.impl.init.registries.deferred.Aspects;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.network.chat.Component;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.inventory.ResultSlot;
import net.minecraft.world.inventory.Slot;
import net.minecraft.world.item.ItemStack;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;
import org.jetbrains.annotations.Nullable;

@OnlyIn(Dist.CLIENT)
public class WorkbenchScreen extends AbstractContainerScreen<WorkbenchMenu> {
  protected static final int PADDING = 2;
  protected static final int SLOT_SIZE = 16;
  private static final Texture WORKBENCH_BG = new Texture("gui/container/arcane_workbench", 216, 134, 255, 255);
  private static final Texture INVENTORY_BG = new Texture("gui/container/inventory", 176, 99, 255, 255);
  private static final Texture SLOT_TEXTURE = new Texture("gui/container/arcane_workbench_slot", 18, 18);
  private static final Texture RESULT_SLOT_TEXTURE = new Texture("gui/container/arcane_workbench_result_slot", 34, 34);
  private static final Texture ASPECT_SLOT_TEXTURE = new Texture("gui/container/arcane_workbench_aspect_slot", 25, 25);

  public WorkbenchScreen(WorkbenchMenu pMenu, Inventory pPlayerInventory, Component pTitle) {
    super(pMenu, pPlayerInventory, pTitle);
    this.inventoryLabelX += 20;
    this.inventoryLabelY += 69;
    this.titleLabelY -= 15;
    this.imageWidth = Math.max(WorkbenchScreen.WORKBENCH_BG.width(), WorkbenchScreen.INVENTORY_BG.width());
    this.imageHeight = WorkbenchScreen.WORKBENCH_BG.height() + WorkbenchScreen.PADDING + WorkbenchScreen.INVENTORY_BG.height();
  }

  @Override
  public void render(GuiGraphics pGuiGraphics, int pMouseX, int pMouseY, float pPartialTick) {
    super.render(pGuiGraphics, pMouseX, pMouseY, pPartialTick);
    this.renderAspects(pGuiGraphics);
    this.renderTooltip(pGuiGraphics, pMouseX, pMouseY);
  }

  @Override
  protected void renderLabels(GuiGraphics guiGraphics, int mouseX, int mouseY) {
    guiGraphics.drawString(this.font, this.title, this.titleLabelX, this.titleLabelY, 0x83FF9B, false);
    guiGraphics.drawString(this.font, this.playerInventoryTitle, this.inventoryLabelX, this.inventoryLabelY, 0x404040, false);
  }

  @Override
  protected void renderBg(GuiGraphics pGuiGraphics, float pPartialTick, int pMouseX, int pMouseY) {
    var t7graphics = new T7GuiGraphics(pGuiGraphics);
    t7graphics.push();
    t7graphics.translateXY(this.leftPos, this.topPos);

    t7graphics.blit(WorkbenchScreen.WORKBENCH_BG);
    t7graphics.translateXY((WorkbenchScreen.WORKBENCH_BG.width() - WorkbenchScreen.INVENTORY_BG.width()) / 2f, WorkbenchScreen.WORKBENCH_BG.height());
    t7graphics.translateXY(0, WorkbenchScreen.PADDING);
    t7graphics.blit(WorkbenchScreen.INVENTORY_BG);

    t7graphics.pop();

    for (var slot : this.menu.slots) this.renderSlotBg(t7graphics, slot);
  }

  @Override
  protected void renderSlotContents(GuiGraphics guiGraphics, ItemStack itemstack, Slot slot, @Nullable String countString) {
    super.renderSlotContents(guiGraphics, itemstack, slot, countString);
    if (slot instanceof ResultSlot resultSlot && !resultSlot.mayPickup(this.menu.getPlayer()))
      guiGraphics.fill(RenderType.guiGhostRecipeOverlay(), slot.x, slot.y, slot.x + 16, slot.y + 16, 0x50FFFFFF);
  }

  protected void renderSlotBg(T7GuiGraphics t7graphics, Slot slot) {
    Texture texture = WorkbenchScreen.SLOT_TEXTURE;
    if (slot instanceof ResultSlot) texture = WorkbenchScreen.RESULT_SLOT_TEXTURE;

    t7graphics.push();
    // go to the center of the slot and draw the texture centered there
    t7graphics.translateXY(this.leftPos + slot.x + WorkbenchScreen.SLOT_SIZE / 2f, this.topPos + slot.y + WorkbenchScreen.SLOT_SIZE / 2f);
    t7graphics.translateXY(-texture.width() / 2f, -texture.height() / 2f);
    t7graphics.blit(texture);
    t7graphics.pop();
  }

  protected void renderAspects(GuiGraphics guiGraphics) {
    final int BASE_RADIUS = 52;
    final float ANGLE = 360f / Aspects.INSTANCE.getPRIMAL_ASPECTS().size();
    var middleSlot = this.menu.slots.get(4);
    var t7graphics = new T7GuiGraphics(guiGraphics);

    t7graphics.push();
    t7graphics.translateXY(this.leftPos, this.topPos);
    t7graphics.translateXY(middleSlot.x, middleSlot.y);

    // draw aspects at hexagon points (or N-gon if more primals are added by addons)
    int i = 0;
    for (var a : Aspects.INSTANCE.getPRIMAL_ASPECTS()) {
      var requiredAmount = this.menu.getRequiredAspects().get(a.get());
      var requiredStack = new AspectStack(a.get(), requiredAmount);
      t7graphics.push();
      t7graphics.rotateZ(ANGLE * i);
      var fac = Math.abs(Math.cos(2*Math.PI / Aspects.INSTANCE.getPRIMAL_ASPECTS().size() * i));
      t7graphics.translateXY((float) (BASE_RADIUS * (1-0.16*fac*fac)), 0);
      t7graphics.rotateZ(-ANGLE * i);
      Texture texture = WorkbenchScreen.ASPECT_SLOT_TEXTURE;
      t7graphics.blit(texture.location(), (SLOT_TEXTURE.width()-ASPECT_SLOT_TEXTURE.width())/2, (SLOT_TEXTURE.height()-ASPECT_SLOT_TEXTURE.height())/2, 0, 0, texture.width(), texture.height(), texture.width(), texture.height());
      if (requiredAmount != 0)
        AspectRenderer.renderAspect(t7graphics, requiredStack, 0, 0);
      t7graphics.pop();
      i++;
    }

    t7graphics.pop();
  }
}

File: src/main/java/me/alegian/thavma/impl/client/screen/WorkbenchScreen.kt
package me.alegian.thavma.impl.client.screen

import me.alegian.thavma.impl.client.T7GuiGraphics
import me.alegian.thavma.impl.client.renderer.AspectRenderer
import me.alegian.thavma.impl.client.texture.Texture
import me.alegian.thavma.impl.common.aspect.AspectStack
import me.alegian.thavma.impl.common.menu.WorkbenchMenu
import me.alegian.thavma.impl.common.menu.slot.DynamicSlot
import me.alegian.thavma.impl.init.registries.deferred.Aspects.PRIMAL_ASPECTS
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.network.chat.Component
import net.minecraft.world.entity.player.Inventory
import kotlin.math.abs
import kotlin.math.cos

private val WORKBENCH_BG = Texture("gui/container/arcane_workbench/bg", 214, 132, 256, 256)
private val RESULT_SLOT = Texture("gui/container/arcane_workbench/result_slot", 34, 34)
private val WAND_SLOT = Texture("gui/container/arcane_workbench/wand_slot", 20, 20)
private val ASPECT_SOCKET = Texture("gui/container/arcane_workbench/aspect_socket", 25, 25)
private val SLOTS = (0..8).map { Texture("gui/container/arcane_workbench/crafting_slots/$it", 21, 21) }
private const val BORDER = 5
private const val WOOD_SIZE = 122

open class WorkbenchScreen(val menu: WorkbenchMenu, pPlayerInventory: Inventory, pTitle: Component) : T7ContainerScreen<WorkbenchMenu>(menu, pPlayerInventory, pTitle, WORKBENCH_BG) {
  override fun ComposeContext.layout() {
    Padding(BORDER) {
      Row {
        Box(Modifier().width(WOOD_SIZE).center()) {
          Box(Modifier().width(SLOTS[0].width * 3).height(SLOTS[0].height * 3).center().color(0xFF0000FF.toInt())) {
            addRenderableOnly(slotGrid(3, 3, menu.craftingContainer.range.slots) { i, j -> SLOTS[i * 3 + j] })
          }
        }
        Box(Modifier().width(BORDER))
        Box {
          Box(Modifier().center()) {
            Box(Modifier().size(RESULT_SLOT)) {
              addRenderableOnly(slot(menu.resultContainer.range.slot, RESULT_SLOT))
            }
          }
          Box(Modifier().maxHeight(0.5f).center()) {
            Box(Modifier().size(WAND_SLOT)) {
              addRenderableOnly(slot(menu.wandContainer.range.slot, WAND_SLOT))
            }
          }
        }
      }
    }
  }

  override fun render(guiGraphics: GuiGraphics, mouseX: Int, mouseY: Int, partialTick: Float) {
    super.render(guiGraphics, mouseX, mouseY, partialTick)
    renderAspects(guiGraphics)
  }

  // TODO: cleanup
  protected open fun renderAspects(guiGraphics: GuiGraphics) {
    val BASE_RADIUS = 56
    val ANGLE = 360f / PRIMAL_ASPECTS.size
    val middleSlot = menu.craftingContainer.range.slots[4]
    if (middleSlot !is DynamicSlot<*>) return
    val t7graphics = T7GuiGraphics(guiGraphics)

    t7graphics.push()
    t7graphics.translateXY(leftPos.toFloat(), topPos.toFloat())
    t7graphics.translateXY(middleSlot.actualX.toFloat(), middleSlot.actualY.toFloat())

    // draw aspects at hexagon points (or N-gon if more primals are added by addons)
    for ((i, a) in PRIMAL_ASPECTS.withIndex()) {
      val requiredAmount = menu.requiredAspects[a.get()]
      val requiredStack = AspectStack(a.get(), requiredAmount)
      t7graphics.push()
      t7graphics.rotateZ(ANGLE * i)
      val fac = abs(cos(2 * Math.PI / PRIMAL_ASPECTS.size * i))
      t7graphics.translateXY((BASE_RADIUS * (1 - 0.16 * fac * fac)).toFloat(), 0f)
      t7graphics.rotateZ(-ANGLE * i)
      t7graphics.blit(ASPECT_SOCKET.location, (SLOTS[0].width - ASPECT_SOCKET.width) / 2, (SLOTS[0].height - ASPECT_SOCKET.height) / 2, 0f, 0f, ASPECT_SOCKET.width, ASPECT_SOCKET.height, ASPECT_SOCKET.width, ASPECT_SOCKET.height)
      if (requiredAmount != 0) AspectRenderer.renderAspect(t7graphics, requiredStack, (SLOTS[0].width - AspectRenderer.getPixelResolution()) / 2, (SLOTS[0].height - AspectRenderer.getPixelResolution()) / 2)
      t7graphics.pop()
    }

    t7graphics.pop()
  }
}